### PR TITLE
Integrate Kubernetes into Edge Agent

### PR DESCRIPTION
**Pull request by kkdawkins**

### CODE REVIEW

**Comment by darobs** [comment]
If you run the "CheckinBuild" target locally, this should catch all the StyleCop errors.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs [hunk 1]
```diff
@@ -360,44 +388,25 @@ private async Task ManageDeployments(V1ServiceList currentServices, V1Deployment
...
                     Events.CreatingDeployment(deployment);
                     return this.client.CreateNamespacedDeploymentAsync(deployment, this.k8sNamespace);
-                });
-            await Task.WhenAll(createDeploymentTasks);
+                }));
 
             // Update the existing - should only do this when different.
-            IEnumerable<Task<V1Deployment>> updateDeploymentTasks = deploymentsUpdated.Select(deployment => this.client.ReplaceNamespacedDeploymentAsync(deployment, deployment.Metadata.Name, this.k8sNamespace));
-            await Task.WhenAll(updateDeploymentTasks);
+            await Task.WhenAll(deploymentsUpdated.Select(deployment => this.client.ReplaceNamespacedDeploymentAsync(deployment, deployment.Metadata.Name, this.k8sNamespace)));
```
Comment (darobs) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs]: Can we wait on all of these tasks at the end? I don't think we have to wait for all the deletes to finish before the new creations and updates - at least for Deployments.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs [hunk 2]
```diff
@@ -309,6 +306,37 @@ private async Task ManageDeployments(V1ServiceList currentServices, V1Deployment
...
+            Dictionary<string, string> currentDeploymentsFromAnnotations = this.GetCurrentDeploymentConfig(currentDeployments)
+                .ToDictionary(
+                    pair => pair.Key,
+                    pair => pair.Value);
+
+            var deploymentsRemoved = new List<V1Deployment>(currentDeployments.Items);
+            deploymentsRemoved.RemoveAll(
+                d =>
+                {
+                    return desiredDeployments.Exists(i => string.Equals(i.Metadata.Name, d.Metadata.Name));
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs]: in this line `i` and `d` make me read the whole function to read multiple times.
```
return desiredDeployments.Exists(desired => desired.Metadata.Name == removing.Metadata.Name);
```
In my opinion, this looks better

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/Program.cs [hunk 3]
```diff
@@ -135,6 +135,54 @@ public static async Task<int> MainAsync(IConfiguration configuration)
                         builder.RegisterModule(new EdgeletModule(iothubHostname, edgeDeviceHostName, deviceId, new Uri(managementUri), new Uri(workloadUri), apiVersion, dockerAuthConfig, upstreamProtocol, proxy, productInfo, closeOnIdleTimeout, idleTimeout));
                         break;
 
+                    case Constants.KubernetesMode:
+                        managementUri = configuration.GetValue<string>(Constants.EdgeletManagementUriVariableName);
```
[the remark this answers is not in the document]
Comment (kkdawkins) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/Program.cs]: Neither IotEdge nor Docker mode do validation here, they do it inside the module. I was keeping with that pattern, but happy to change it.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/KubernetesModule.cs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            // IKubernetesClient
+            builder.Register(c =>
+            {
+                if (this.enableServiceCallTracing)
+                {
+                    // enable tracing of k8s requests made by the client
+                    var loggerFactory = c.Resolve<ILoggerFactory>();
+                    ILogger logger = loggerFactory.CreateLogger(typeof(Kubernetes));
+                    ServiceClientTracing.IsEnabled = true;
+                    ServiceClientTracing.AddTracingInterceptor(new DebugTracer(logger));
```
[the remark this answers is not in the document]
Comment (kkdawkins) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/KubernetesModule.cs]: It is :) look at line 96 
`if (this.enableServiceCallTracing)`

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/Program.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/Program.cs]: Yes, the KubernetesModule and AgentModule do all the validation. We probably don't need it here, too.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/KubernetesModule.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkdawkins) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/KubernetesModule.cs]: @dmolokanov  Why the confused emoji?

**Pull request merged**